File: clients/csharp/Model/Player.cs
namespace SpbAiChamp.Model
{
    /// <summary>
    /// Player (game participant)
    /// </summary>
    public struct Player
    {
        /// <summary>
        /// Team index
        /// </summary>
        public int TeamIndex { get; set; }
        /// <summary>
        /// Current score points
        /// </summary>
        public int Score { get; set; }
        /// <summary>
        /// Player's specialty
        /// </summary>
        public Model.Specialty? Specialty { get; set; }
    
        public Player(int teamIndex, int score, Model.Specialty? specialty)
        {
            this.TeamIndex = teamIndex;
            this.Score = score;
            this.Specialty = specialty;
        }
    
        /// <summary> Read Player from reader </summary>
        public static Player ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new Player();
            result.TeamIndex = reader.ReadInt32();
            result.Score = reader.ReadInt32();
            if (reader.ReadBoolean())
            {
                result.Specialty = SpecialtyHelper.ReadFrom(reader);
            } else
            {
                result.Specialty = null;
            }
            return result;
        }
    
        /// <summary> Write Player to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(TeamIndex);
            writer.Write(Score);
            if (!Specialty.HasValue)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                writer.Write((int) (Specialty.Value));
            }
        }
    
        /// <summary> Get string representation of Player </summary>
        public override string ToString() {
            string stringResult = "Player { ";
            stringResult += "TeamIndex: ";
            stringResult += TeamIndex.ToString();
            stringResult += ", ";
            stringResult += "Score: ";
            stringResult += Score.ToString();
            stringResult += ", ";
            stringResult += "Specialty: ";
            if (!Specialty.HasValue)
            {
                stringResult += "null";
            } else
            {
                stringResult += Specialty.Value.ToString();
            }
            stringResult += " }";
            return stringResult;
        }
    }
}

File: clients/go/model/player.go
package model

import "fmt"
import "io"
import . "spb_ai_champ/stream"

// Player (game participant)
type Player struct {
    // Team index
    TeamIndex int32
    // Current score points
    Score int32
    // Player's specialty
    Specialty *Specialty
}

func NewPlayer(teamIndex int32, score int32, specialty *Specialty) Player {
    return Player {
        TeamIndex: teamIndex,
        Score: score,
        Specialty: specialty,
    }
}

// Read Player from reader
func ReadPlayer(reader io.Reader) Player {
    var teamIndex int32
    teamIndex = ReadInt32(reader)
    var score int32
    score = ReadInt32(reader)
    var specialty *Specialty
    if ReadBool(reader) {
        var specialtyValue Specialty
        specialtyValue = ReadSpecialty(reader)
        specialty = &specialtyValue
    } else {
        specialty = nil
    }
    return Player {
        TeamIndex: teamIndex,
        Score: score,
        Specialty: specialty,
    }
}

// Write Player to writer
func (player Player) Write(writer io.Writer) {
    teamIndex := player.TeamIndex
    WriteInt32(writer, teamIndex)
    score := player.Score
    WriteInt32(writer, score)
    specialty := player.Specialty
    if specialty == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        specialtyValue := *specialty
        WriteInt32(writer, int32(specialtyValue))
    }
}

// Get string representation of Player
func (player Player) String() string {
    stringResult := "{ "
    stringResult += "TeamIndex: "
    teamIndex := player.TeamIndex
    stringResult += fmt.Sprint(teamIndex)
    stringResult += ", "
    stringResult += "Score: "
    score := player.Score
    stringResult += fmt.Sprint(score)
    stringResult += ", "
    stringResult += "Specialty: "
    specialty := player.Specialty
    if specialty == nil {
        stringResult += "nil"
    } else {
        specialtyValue := *specialty
        stringResult += SpecialtyToString(specialtyValue)
    }
    stringResult += " }"
    return stringResult
}